lib/shared/drive-constraints.js

```diff
@@ -208,3 +208,162 @@ exports.isDriveValid = (drive, image) => {
+   *
+   * @description
+   * The drive is too small for the image.
+   */
+  TOO_SMALL: 'Too Small For Image',
+
+  /**
+   * @property {String} LOCKED
+   * @memberof COMPATIBILITY_STATUS_MESSAGES
+   *
+   * @description
+   * The drive is locked (SD cards) and cannot be written to.
+   */
+  LOCKED: 'Locked',
+
+  /**
+   * @property {String} SYSTEM
+   * @memberof COMPATIBILITY_STATUS_MESSAGES
+   *
+   * @description
+   * The drive is a system drive and should not be written to.
+   */
+  SYSTEM: 'System Drive',
+
+  /**
+   * @property {String} CONTAINS_IMAGE
+   * @memberof COMPATIBILITY_STATUS_MESSAGES
+   *
+   * @description
+   * The drive contains the image and therefore cannot be written to.
+   */
+  CONTAINS_IMAGE: 'Drive Contains Image'
+};
+
+/**
+ * @summary Drive/image compatibility status types.
+ * @public
+ * @type {Object}
+ *
+ * @description
+ * Status types classifying what kind of message it is, i.e. ok, error, warning.
+ */
+exports.COMPATIBILITY_STATUS_TYPES = {
+  OK: 0,
+  ERROR: 1,
+  WARNING: 2
+};
+
+/**
+ * @summary Get drive/image compatibility in an object
+ * @function
+ * @public
+ *
+ * @description
+ * Given an image and a drive, return their compatibility status object
+ * containing the status type (ERROR, WARNING, and OK), and accompanying
+ * status message.
+ *
+ * @param {Object} drive - drive
+ * @param {Object} image - image
+ * @returns {Object} compatibility status object
+ *
+ * @example
+ * const drive = {
+ *   device: '/dev/disk2',
+ *   name: 'My Drive',
+ *   size: 4000000000
+ * };
+ *
+ * const image = {
+ *   path: 'rpi.img',
+ *   size: 2000000000
+ *   recommendedDriveSize: 4000000000
+ * });
+ *
+ * const { type, message } = constraints.getDriveImageCompatibilityStatus(drive, image);
+ *
+ * if (type === constraints.COMPATIBILITY_STATUS_TYPES.WARNING) {
+ *   // do something
+ * } else if (type === constraints.COMPATIBILITY_STATUS_TYPES.ERROR) {
+ *   // do something else
+ * }
+ */
+exports.getDriveImageCompatibilityStatus = (drive, image) => {
+
+  // Mind the order of the if-statements if you modify.
+  if (exports.isSourceDrive(drive, image)) {
+    return {
+      type: exports.COMPATIBILITY_STATUS_TYPES.ERROR,
+      message: exports.COMPATIBILITY_STATUS_MESSAGES.CONTAINS_IMAGE
+    };
+  }
+
+  if (exports.isSystemDrive(drive)) {
+    return {
+      type: exports.COMPATIBILITY_STATUS_TYPES.WARNING,
+      message: exports.COMPATIBILITY_STATUS_MESSAGES.SYSTEM
+    };
+  }
+
+  if (!_.isNil(drive) && !exports.isDriveLargeEnough(drive, image)) {
+    return {
+      type: exports.COMPATIBILITY_STATUS_TYPES.ERROR,
+      message: exports.COMPATIBILITY_STATUS_MESSAGES.TOO_SMALL
+    };
+  }
+
+  if (exports.isDriveLocked(drive)) {
+    return {
+      type: exports.COMPATIBILITY_STATUS_TYPES.ERROR,
+      message: exports.COMPATIBILITY_STATUS_MESSAGES.LOCKED
+    };
+  }
+
+  if (!_.isNil(drive) && !exports.isDriveSizeRecommended(drive, image)) {
```

Comment:
Hmmm, based on my comment above, this is a tricky one. If the drive is both a system drive _AND_ smaller than the image's recommended size, which warning should we display? From a logical point of view, the user really ought to be made aware of _both_ warnings, because they both signify different things.
If we only display the "system" warning, then the user isn't aware that the drive is also smaller than the image's recommended size.
If we display only the "not recommended" warning, then the user isn't aware that they'll potentially be overwriting one of their system drives.
:-S

(but given the choice of only displaying one warning, the "system" warning obviously has higher priority)